tools/crm_resource_runtime.c

```diff
@@ -1290,14 +1290,35 @@ cli_resource_restart(pcmk__output_t *out, pe_resource_t *rsc, const char *host,
 
     pe_working_set_t *data_set = NULL;
 
-    if (!resource_is_running_on(rsc, host)) {
-        const char *id = rsc->clone_name?rsc->clone_name:rsc->id;
-        if(host) {
-            out->err(out, "%s is not running on %s and so cannot be restarted", id, host);
+    /* If host is NULL, there's no need to check anything.  A NULL host will
+     * match all hosts.
+     */
+    if (host != NULL) {
+        bool found = false;
+
+        if (pe_rsc_is_clone(rsc->parent)) {
+            for (GList *iter = uber_parent(rsc)->children; iter != NULL; iter = iter->next) {
+                pe_resource_t *child = iter->data;
+
+                if (resource_is_running_on(child, host)) {
+                    rsc = child;
+                    found = true;
+                    break;
+                }
+            }
         } else {
-            out->err(out, "%s is not running anywhere and so cannot be restarted", id);
+            found = resource_is_running_on(rsc, host);
+        }
```

Comment:
This makes me wonder about resource_is_running_on() itself. At a minimum, its node name comparison should be case-insensitive.

I'm not sure whether your change here should be in resource_is_running_on() instead, but I suspect it should. Looking at the other callers:

* get_active_resources(): at first glance, this is likely unaffected by any changes since it only iterates through the cluster's top-level resources and group members. However I notice it expands groups, but not cloned groups, which may be a problem.
* waiting_for_starts(): probably needs the same fix
* resource_reasons_list_default() and resource_reasons_list_xml(): probably needs same fix

Reply:
I've gone ahead and added the case-insensitive bit.